### PR TITLE
Bin TQ MVA output

### PR DESCRIPTION
**Pull request by cgsavard**
#### PR description:

This implements the binning scheme for the output of the TQ MVA1 variable. This binning scheme was presented at the offline software general meeting, [here](https://indico.cern.ch/event/1401028/). In order to avoid applying a logistic sigmoid function (which puts the variable within a 0-1 range) to the BDT output in the firmware, the MVA variable is binned before the transformation is applied and the normal calling of trkMVA1() in the TTTrack class will apply the transformation. An additional call to a pre-transformed variable was also included for convenience and extra checks.

#### PR validation:

All checks were run and this was tested with both the HYBRID and NEW_KF configurations of the tracking code. Here are some outputs showing the conversion between the different MVA variables in order to check that everything is working as expected (bins are [-480, -2.197, -1.386, -0.405, 0.405, 1.386, 2.197, 2.944] for pre-logistic sigmoid, corresponding to [0, 0.1, 0.2, 0.4, 0.6, 0.8, 0.9, 0.95]  after the transform):

HYBRID tracking example 1:
TTTrack->trkMVA1Pre(): 2.31981	
TTTrack->trkMVA1(): 0.910505	
TTTrack_Word->getMVAQualityBits(): 6	
TTTrack_Word->getMVAQualityPre(): 2.197	
TTTrack_Word->getMVAQuality(): 0.9

HYBRID tracking example 2:
TTTrack->trkMVA1Pre(): 2.01361	
TTTrack->trkMVA1(): 0.882219	
TTTrack_Word->getMVAQualityBits(): 5	
TTTrack_Word->getMVAQualityPre(): 1.386	
TTTrack_Word->getMVAQuality(): 0.8

NEWKF tracking example (default set to 0, needs fix in separate PR):
TTTrack->trkMVA1Pre(): -480	
TTTrack->trkMVA1(): 3.4566e-209	
TTTrack_Word->getMVAQualityBits(): 0	
TTTrack_Word->getMVAQualityPre(): -480	
TTTrack_Word->getMVAQuality(): 0

### CODE REVIEW

**Comment by tomalin** [comment]
Hi @cgsavard, my only worry here is what happens if the L1trigger group has already produced a MC containing all the L1 trigger objects, such as L1tracks, and they try reading it with a newer version of CMSSW that contains this PR. I don't know if CMSSW would warn them of their mistake by throwing a fatal error because of the name change of the data member TTTrack::theTrkMVA1? Or would no error be thrown, and the old theTrkMVA1 value would be loaded into the theTrkMVA1pre data member, giving nonsense.
So do we have to release this with a major CMSSW version change (e.g. CMSSW 15), accompanied by warnings to the trigger group?

**Comment by tomalin** [comment]
Minor comment. adding "pre" to variable name "theMVA1pre" is a little obscure. But I don't have a better suggestion.

**Comment by tomalin** [comment]
If you think it's likely that MVA2 & MVA3 will need a similar approach, then I suggest you make that change now, at least for the data members, so that the data members of this TTTrack EDProduct are not changing too often, causing bcackwards incompatbility.

**Comment by cgsavard** [comment]
@tomalin I'm not sure I follow entirely the TTTrack::theTrkMVA1 discussion. Are you talking about the switch [here](https://github.com/cms-L1TK/cmssw/pull/272/files#diff-a5408174f85949904e85b6a188ef6e5eefaa0da1cf3b6f11eaf2e7e410ead4daL48-R48)? I'm not sure this switch would affect anything since they are private variables so they are only recognized internally in the class. Unless for some reason the L1Trigger group can access this? I imagine their samples use the [TTTrack<T>::trkMVA1()](https://github.com/cms-L1TK/cmssw/blob/14fdfe10f1bf82d85d7cc35238ca95e1c9603926/DataFormats/L1TrackTrigger/interface/TTTrack.h#L384) call and this function represents the same value so no mistmatch will show up there. Perhaps I am mistunderstanding how the L1Trigger MC samples can use the private theTrkMVA1 variables?

I imagine we might want to use MVA2 for a displaced TQ BDT, which I've been getting a lot of inquiries about people wanting to try out. I imagine that that BDT would require the same set up, so I can make that change. As for MVA3, there hasn't really been any discussion on what this may be set with. If it's not a BDT, then it may require a slightly different set up. If in the above paragraph we decide the change from theTrkMVA1->theTrkMVAPre1 won't change anything then I am in favor of keeping both of these as is and I will change them when we decide how to fill MVA2 and MVA3. Otherwise, I can go ahead and make the change for MVA2 now but am unsure about MVA3. What do you think is best?

**Comment by tomalin** [comment]
> @tomalin I'm not sure I follow entirely the TTTrack::theTrkMVA1 discussion. Are you talking about the switch [here](https://github.com/cms-L1TK/cmssw/pull/272/files#diff-a5408174f85949904e85b6a188ef6e5eefaa0da1cf3b6f11eaf2e7e410ead4daL48-R48)? I'm not sure this switch would affect anything since they are private variables so they are only recognized internally in the class. Unless for some reason the L1Trigger group can access this? I imagine their samples use the [TTTrack::trkMVA1()](https://github.com/cms-L1TK/cmssw/blob/14fdfe10f1bf82d85d7cc35238ca95e1c9603926/DataFormats/L1TrackTrigger/interface/TTTrack.h#L384) call and this function represents the same value so no mistmatch will show up there. Perhaps I am mistunderstanding how the L1Trigger MC samples can use the private theTrkMVA1 variables?
> 
> I imagine we might want to use MVA2 for a displaced TQ BDT, which I've been getting a lot of inquiries about people wanting to try out. I imagine that that BDT would require the same set up, so I can make that change. As for MVA3, there hasn't really been any discussion on what this may be set with. If it's not a BDT, then it may require a slightly different set up. If in the above paragraph we decide the change from theTrkMVA1->theTrkMVAPre1 won't change anything then I am in favor of keeping both of these as is and I will change them when we decide how to fill MVA2 and MVA3. Otherwise, I can go ahead and make the change for MVA2 now but am unsure about MVA3. What do you think is best?

Hi Claire, if an existing MC dataset contains TTTrack objects made prior to this PR, then the data members of this class, which is what is stored in the .root file, include theTrkMVA1_ . If this file is read into a CMSSW version that includes the code changes in this PR, then and someone's C++ requests (via Handle) to load TTTrack from disk into memory, then ROOT will be confused the variable theTrkMVA1 no longer exists in the TTTrack objects that it' being asked to fill in memory. It might guess that it can load the value from theTrkMVA1 into theTrkMVA1pre, but there's a good chance it will mess up. This is nothing to do with private or public functions or data members.

**Comment by cgsavard** [comment]
@tomalin Do you have a solution in mind that I can implement?

**Comment by tomalin** [comment]
> @tomalin Do you have a solution in mind that I can implement?

Hi Claire, one thing is to minimise how often we change the data members of EDProducts. Hence the suggestion to change MVA2 & MVA3 at  the same time. The other is time time the PR for a major CMSSW release, but CMSSW 15 not foreseen until late 2024 https://twiki.cern.ch/twiki/bin/viewauth/CMS/ReleaseSchedule. I suggest you ask CMS SW group for advice.

**Comment by Chriisbrown** [comment]
Hi having had a close look at this I realise that changes have already been made to the track word to include the track quality binning and associated functions in a similar manner to the chi2 variables. This has been the case in central cmssw since CMSSW14_1_0_pre2 (released march 27) (https://github.com/cms-sw/cmssw/blob/master/DataFormats/L1TrackTrigger/interface/TTTrack_TrackWord.h#L106). So once we rebase our dev branch here to that we'll bring in those changes, so Claire's changes to the track word will cause merge conflicts and then possibly break any downstream L1 trigger code.

**Comment by cgsavard** [comment]
@tomalin so the SW group said that the default behavior is that ```trk_MVA_``` will be ignored and ```trk_MVA1Pre_``` will be set to 0 (which means that the trk_MVA1() function will give 0.5). And I think this is fine for the transition as the only people that use the MVA variables that I know so far is the GTT group and I plan to make these same changes in their repo and will make them aware of the change. I suggest making the same change to the other two MVA variables, and then opening the same PR in l1t-offline.

For the point @Chriisbrown made, I suggest also opening a PR into central CMSSW making the proper changes there, that way everything will match. I see that this MVA variable is used in the NNVtx code so I'll also make sure that that variable called is the proper one.

How does this all sound?

File: L1Trigger/TrackFindingTracklet/plugins/L1FPGATrackProducer.cc
Comment (cgsavard) [inline L1Trigger/TrackFindingTracklet/plugins/L1FPGATrackProducer.cc:768]: I actually realize that moving this after setting the track quality in the TTTrack is not correct. The L1 track quality pulls some [inputs from the track word](https://github.com/cms-L1TK/cmssw/blob/L1TK-dev-14_0_0_pre2/L1Trigger/TrackTrigger/src/L1TrackQuality.cc#L66-L68) which are just set to 0 if the track word isn't set before running the TQ MVA. I think we should move the setting of the track word to before setting L1 track quality, and then just set the track word MVA after. This would mean including another set function in TTTrack_Word just for MVA. And we should also consider swapping all the input variables to the track word version instead of the TTTrack versions. Thoughts @tomalin ?

Comment (cgsavard) [reply L1Trigger/TrackFindingTracklet/plugins/L1FPGATrackProducer.cc:768]: Alternatively, we run ```aTrack.setTrackWordBits();``` before and after MVA is set in the L1 track, but that's maybe not super nice.

**Comment by tomalin** [comment]
> @tomalin so the SW group said that the default behavior is that `trk_MVA_` will be ignored and `trk_MVA1Pre_` will be set to 0 (which means that the trk_MVA1() function will give 0.5). And I think this is fine for the transition as the only people that use the MVA variables that I know so far is the GTT group and I plan to make these same changes in their repo and will make them aware of the change. I suggest making the same change to the other two MVA variables, and then opening the same PR in l1t-offline.
> 
> For the point @Chriisbrown made, I suggest also opening a PR into central CMSSW making the proper changes there, that way everything will match. I see that this MVA variable is used in the NNVtx code so I'll also make sure that that variable called is the proper one.
> 
> How does this all sound?

@cgsavard This sounds like a good plan. Though please tell Keith Ulmer and the L1 trigger group conveners about this (when your PR is ready), so that it doesn't catch anyone by surprise.

**Comment by Chriisbrown** [comment]
@cgsavard from the point of view of the NNvtx any changes to the track word including a rebinning of the MVA output would require us to do a full retrain and given the NNvtx has only just gone into cmssw and any update to the model then takes a separate PR to cms-data this would take time. Do we (cms-l1tk) need to do this change to the track word now or can we bring in the changes from central cmssw and use that implementation of the MVA in the track word? And then at some point in the future make this update to the track word after testing any update to the binning scheme with the downstream users such as the NNvtx. Either way Benjamin requests that it is discussed in a GTT meeting

File: L1Trigger/TrackFindingTracklet/plugins/L1FPGATrackProducer.cc
Comment (tomalin) [reply L1Trigger/TrackFindingTracklet/plugins/L1FPGATrackProducer.cc:768]: I don't see a pretty solution.

**Comment by tomalin** [comment]
> @cgsavard from the point of view of the NNvtx any changes to the track word including a rebinning of the MVA output would require us to do a full retrain and given the NNvtx has only just gone into cmssw and any update to the model then takes a separate PR to cms-data this would take time. Do we (cms-l1tk) need to do this change to the track word now or can we bring in the changes from central cmssw and use that implementation of the MVA in the track word? And then at some point in the future make this update to the track word after testing any update to the binning scheme with the downstream users such as the NNvtx. Either way Benjamin requests that it is discussed in a GTT meeting

Hi Chris, the code you put in central CMSSW used different MVA binning https://github.com/cms-sw/cmssw/blob/master/DataFormats/L1TrackTrigger/interface/TTTrack_TrackWord.h#L106 to Claire's new code https://github.com/cms-L1TK/cmssw/blob/tq_binning_reorganize/DataFormats/L1TrackTrigger/interface/TTTrack_TrackWord.h#L107 . If she makes her PR, but retains your binning (until some future agreed time when it's retuned), would that solve your problem? (I assume the change to pre-sigmoid binning is no problem?).

**Comment by Chriisbrown** [comment]
Hi Ian,
This would be fine so long as the getMVA1() functions and the MVA bits in the track word give the same answer between the PRs. Though thinking about this when we tune the BDT output it'll be done in the 0-1 space, it isn't clear how we would create the "presigmoid" bins that the trackword now wants and ensure that we are doing is consistent every time

**Comment by cgsavard** [comment]
@Chriisbrown The getMVA1() function and bits will be exactly the same as what you are using if you just change the bins to what you are using. But we should definitely get together and finalize the bins as there are other GTT objects that also use the MVA and I partially tuned the bins with those objects in mind (track jets, track HT, track MET). Regardless of the binning scheme, this is the proper way we should be binning everything since we do not want to do a logistic sigmoid on the firmware. And again, it will not affect your NNvtx as long as you change the binning to what you want.

**Comment by Chriisbrown** [comment]
@cgsavard I don't know if just changing the binning is a viable solution when using the NNvtx as the upcoming MC campaign in CMSSW 14 will use the previous binning meaning the tracks created in that campaign will have that binning. If we were to change the binning now anyone rerunning the track finding will then have to adapt the binning if they want to see the same MVA track word which has the potential to confuse users of the NNvtx or the BDT in general, if we are going to change the binning we need to time it with a major CMSSW release or a major MC campaign with an accompanying talk in a L1 trigger meeting of some sort. I think we should have understandable bins in software (between 0 and 1) and derive the presigmoid bins for the FW, any changes to the binning will be done by a software developer where the 0 to 1 scheme makes sense not a firmware developer where the bins are just some constants in a file (https://github.com/cms-L1TK/l1tk-for-emp/blob/47e69eb20fd7573644a3c80c74dee96b49f8ab9e/kfout/firmware/hdl/kfout_config.vhd#L61C10-L61C22)

**Comment by cgsavard** [comment]
@Chriisbrown So are you suggesting we have two sets of bins in software, one with the 0-1 binning scheme and one with the pre-sigmoid? Right now we've got two different places that are being binned and each of these places needs a different binning scheme. We can certainly bin the track word value post-sigmoid, but you also set up the KFOut module to bin the tq variable pre-sigmoid [here](https://github.com/cms-l1t-offline/cmssw/blob/phase2-l1t-integration-14_0_0_pre3/L1Trigger/TrackFindingTracklet/plugins/ProducerKFout.cc#L301) to match the firmware and that means we also need pre-sigmoid bins. I don't see any issue with the next person who studies and changes the bins to just apply an inverse sigmoid when setting them, and the MVA users will still get the 0-1 binning scheme when using it downstream because I set up the track word class to apply this transformation [here](https://github.com/cms-L1TK/cmssw/blob/14fdfe10f1bf82d85d7cc35238ca95e1c9603926/DataFormats/L1TrackTrigger/interface/TTTrack_TrackWord.h#L234).

Once again, nothing has changed from the user's side. trkMVA1() still gives the track float on a 0-1 range and getMVAQuality() still gives the digitized 0-1 values from the track word. The only reason the bins are set pre-sigmoid is because in the KFout module we are binning the pre-sigmoid value. If we want to bin the post-sigmoid value in the KFout then I think we can switch it all back and make that change, but I thought you set it up that way specifically to match the firmware. What we definitely shouldn't do is have two different places where bins are set like what is done now [here](https://github.com/cms-l1t-offline/cmssw/blob/phase2-l1t-integration-14_0_0_pre3/DataFormats/L1TrackTrigger/interface/TTTrack_TrackWord.h#L106) and [here](https://github.com/cms-l1t-offline/cmssw/blob/phase2-l1t-integration-14_0_0_pre3/L1Trigger/TrackTrigger/python/TrackQualityParams_cfi.py#L21) because that's messy and bound to cause mistakes in the future.

We can definitely change the bins I have set here to what you guys want for the NNVtx for now, but we will have to discuss another update to this further as many of the other GTT objects are affected by this.

**Comment by Chriisbrown** [comment]
No I think we should have one sets of bins in the track word only and those bins should be between 0 and 1 so that when a downstream user tunes their threshold between 0 and 1 they can know if this makes sense in the binning. The KFout can calculate what bins it needs itself and either have a way of printing them for the firmware or the firmware can derive them, but this can be put in place in the large rewrite of the ProducerKFout and ProducerTT that will be done.  The reason the KFout does it the way it does now is because there was no binning in the track word and I wanted to avoid any unnecessary updates to the dataformats, having the bins as a python config also allowed for less disruptive testing of different binning but now the BDT is in more widespread use it makes sense to have a binning in place in the track word and for the KFout to pull the binning from there.

**Comment by cgsavard** [comment]
So am I ok to add a logistic sigmoid to the KFout so that I can use the same bins? Or is there a specific reason that it is not already done and instead I should use an inverse sigmoid on the 0-1 bin value? I want to avoid using an inverse sigmoid since it's not defined at 0.

**Comment by Chriisbrown** [comment]
I would not add a sigmoid to the output of the MVA in KFout as this isn't what happens in FW, instead I would replace tqBins_    with an array that is derived from the (0-1) bins defined in the track word, so performing an inverse sigmoid on those bins.  You can set the highest and lowest bins (the 0 and 1 that are undefined in the inverse sigmoid) as the maximum and minimum possible values the BDT can produce which is the maximum number that can be represented by an ap_fixed<10,5> and -1*that (this is +/- 16 for the ap_fixed<10,5> case but you may want to make this generic) . In the KFout there is an ap_fixed rescale that takes you from ap_fixed to integer hence the multiplication before the binning, this is to match the FW so whatever bins you create will need also to be multiplied by that ap_fixed_rescale before binning. As the KFout currently only produces tracks for the demonstrator tools this can be done now or when the KFout is updated to also produce the TTTrackwords

**Comment by cgsavard** [comment]
This will work now with how the KFout is set up, but eventually when the KFout sets the track word this becomes tricky. Do you imagine the track word being set with the tq bins or with the output tq variable? If it's with the latter then we will have to do an inverse transform on the MVA output and that can give us something undefined.

Perhaps there is another solution that makes us all happy. What if everything is done as I have set it up, but make the binning set up on the 0-1 range and there is an automatic conversion that happens to the pre-sigmoid binning within the track word class. This way the track and track word are still initialized pre-sigmoid, the functions to get the MVA variables are still on the 0-1 range, we avoid any sort of inverse transform in the future, our software matches what is done in firmware, and the user sees and sets a 0-1 binning which will make more sense to them.

**Comment by Chriisbrown** [comment]
I don't know if we want to restrict ourselves to a sigmoid activation function on the end of whatever track quality algorithm we train or design, the only reason there is this situation with the presigmoided bins is a optimisation conifer makes on the output of the BDT in firmware, we could include the sigmoid in the firmware but it would be a waste of resources. If we were to replace the BDT with an NN or BDT not trained on binary cross entropy we would have to rewrite the track word class.

A solution that I think removes this is putting a sigmoid in the KFout to only set the floating point trkMVA1. We run the emulated BDT in the KFout, this produces an “unsigmoided” value, this is then binned in the “unsigmoided” bins that we derive from the 0-1 bins in the trackword definition to give an integer between 0 and 7. That is put in the 3-bits of the TTTrackword. In fact we can set the entire TTTrack word here with the bits that the KFout produces, which is exactly what happens in firmware.

We can also sigmoid the output of the BDT to produce a value between 0 and 1 and this is then set as the trkMVA1 for the floating point tttrack. If we want we can check to see if the same bits are produced by the binning of the “presigmoid BDT” in its “presigmoid” bins and the binning of the “sigmoid BDT” and its 0-1 bins are the same, we know every possible value the BDT can produce (between -16 and 16 in 1024 steps) so can quickly scan across these. 

What this does open up is the possibility that someone runs a setttrackwordbits after the TrackWord has been set by us in the KFout and the TrackWord changes as a result. Though anyone downstream changing bits in the Track Word that we set should know that this may give them incorrect results so I don’t think this is a problem. This way our firmware and emulation produce exactly the same tttrack words that downstream can use and we can tune the BDT binning based off studies in the 0-1 range.

**Comment by cgsavard** [comment]
So I changed the bins to the 0-1 range and changes the naming scheme back to reflect the post-sigmoid values. And the binning scheme is what is implemented in the CMSSW. @tomalin @Chriisbrown Does this look ok?

File: L1Trigger/TrackerTFP/src/DataFormats.cc
Comment (Chriisbrown) [inline L1Trigger/TrackerTFP/src/DataFormats.cc]: This should be set to 0. now

File: L1Trigger/TrackFindingTracklet/plugins/ProducerKFout.cc
Comment (Chriisbrown) [inline L1Trigger/TrackFindingTracklet/plugins/ProducerKFout.cc]: Check that the ap_fixed_rescale multiplication isn't changing what's going on here, I think if you want to put a sigmoid on the output and not a inverse sigmoid on the bins you need to get rid of the ap_fixed_rescale multiplication but then we aren't matching emulation, this might not matter if the bin that is selected is the same anyway. The inverse sigmoid on the bins for defining this TTBV tqMVA might be better and then whenever we set the floating point trkMVA1 use a sigmoid on the output of the BDT.

File: L1Trigger/TrackTrigger/python/TrackQualityParams_cfi.py
Comment (Chriisbrown) [inline L1Trigger/TrackTrigger/python/TrackQualityParams_cfi.py]: As we now use a .json for the model do we want to change the naming convention throughout to move away from "ONNXmodel" to prevent confusion.  We don't actually call the ONNX runtime anywhere so do we still need to include it at the top of the L1TrackQuality.h?

File: L1Trigger/TrackFindingTracklet/plugins/ProducerKFout.cc
Comment (cgsavard) [reply L1Trigger/TrackFindingTracklet/plugins/ProducerKFout.cc]: I implemented a function in TTTrack_Word to convert the 0-1 bins to the pre-sigmoid ones, and after testing it (and fixing a bug in the digitise function), I'm not sure the ap_fixed_rescale is working as expected. Here is a list of outputs for a couple of tracks:
```
post-sig value 0.992879, pre-sig value 4.9375, post-sig bin 7, pre-sig bin 7, pre-sig bin with rescale, 7
post-sig value 0.492188, pre-sig value -0.03125, post-sig bin 3, pre-sig bin 3, pre-sig bin with rescale, 2
post-sig value 0.972415, pre-sig value 3.5625, post-sig bin 7, pre-sig bin 7, pre-sig bin with rescale, 7
post-sig value 0.893309, pre-sig value 2.125, post-sig bin 7, pre-sig bin 7, pre-sig bin with rescale, 7
post-sig value 0.924142, pre-sig value 2.5, post-sig bin 7, pre-sig bin 7, pre-sig bin with rescale, 7
post-sig value 0.989672, pre-sig value 4.5625, post-sig bin 7, pre-sig bin 7, pre-sig bin with rescale, 7
post-sig value 0.384912, pre-sig value -0.46875, post-sig bin 3, pre-sig bin 3, pre-sig bin with rescale, 0
```
We can see the post-sig and pre-sig bins all match and are doing what is expected, and the rescale is multiplying the pre-sig value by 32.0 which sends the new value into the wrong bin when the pre-sig value is closer to 0. Why is this rescale happening? It seems to be causing quite a few binning mistakes and I don't think we would ever get anything within the 3-5 bins because of it.

File: L1Trigger/TrackTrigger/python/TrackQualityParams_cfi.py
Comment (cgsavard) [reply L1Trigger/TrackTrigger/python/TrackQualityParams_cfi.py]: Good idea, I'll change the naming convention.

File: L1Trigger/TrackFindingTracklet/plugins/ProducerKFout.cc
Comment (Chriisbrown) [reply L1Trigger/TrackFindingTracklet/plugins/ProducerKFout.cc]: What bins do you have for the presigmoid case? What was the bug is the digitise function? Because when I do this on paper I get the same bins as I the post-sig version? Do you multiply your bins by the same ap_fixed rescale? The rescale is done because in the emulation we use the ap_fixed<10,5> datatype, in the firmware we use integers so to get from the ap_fixed type to integer we multiply by 2**(10 - 5) = 32. The result should be the same without the ap_fixed_rescale so long as the bins you are using are also without the ap_fixed_rescale but it makes comparing to the FW a lot easier.

Comment (Chriisbrown) [reply L1Trigger/TrackFindingTracklet/plugins/ProducerKFout.cc]: The current code that this comment points to will not work with the ap_fixed_rescale as it comes before the sigmoid function. We need to use presigmoided bins, multiply them by the ap_fixed_rescale and then use them to digitise the output of the BDT after it has been multiplied by the ap_fixed_rescale. Though it shouldn't matter if the rescale is there or not it is just there so we can compare to FW more easily

Comment (cgsavard) [reply L1Trigger/TrackFindingTracklet/plugins/ProducerKFout.cc]: Ok so the issue must be that the pre-sigmoided bins were not rescaled. I will remove the rescaling as it seems a bit silly to multiply both the value we are comparing against and the bins by the same value, and that scaling can just be done easily when comparing with firmware. But I will keep the pre-sigmoided binning in the KFout. The bug in the digitise function was that the last bin is never being checked since the last bin is technically the last value in the array out to infinity. For the chi2 variables this bin is rarely filled but for the MVA the last bin is very often filled so perhaps that's why it was never caught before. Anyways, I'll make the changes and resolve this and then I think we should be good to go.

**Comment by cgsavard** [comment]
All comments have been resolved and this should be ready to merge.

File: DataFormats/L1TrackTrigger/interface/TTTrack_TrackWord.h
Comment (Chriisbrown) [inline DataFormats/L1TrackTrigger/interface/TTTrack_TrackWord.h]: This should be -16 at least for the ap_fixed<10,5> case, though is entirely dependent on what digitisation you have chosen for the BDT so is a hard coded solution. Another reason for this not to be in the track word and instead derived from the precision of the BDT in the KF out.

File: DataFormats/L1TrackTrigger/interface/TTTrack_TrackWord.h
Comment (Chriisbrown) [inline DataFormats/L1TrackTrigger/interface/TTTrack_TrackWord.h:1]: I don't think we should be hardcoding the sigmoid activation into the track word, this should be part of the KFout where the BDT is called or part of the track quality package. All the other parameters in the track word have no additional processing like this.

Comment (cgsavard) [reply DataFormats/L1TrackTrigger/interface/TTTrack_TrackWord.h:1]: Well it depends on whether we think we will use the emulation version of the BDT in the simulation as well, which I think we were saying with Ian we may switch to to match the emulation and firmware. So from this perspective, and for debugging, I think it makes sense to have the converted bins in the TTTrack_Word so that it can be called anywhere. And this is also why I think we shouldn't switch to -16 for the first bin because the simulation might produce results below that. Actually now that I think about this, because in the emulation version of the processing where <10,5> goes in and out, will the lowest result always be -15? Because then I think it makes sense to switch to -15.

Anyways, I suggest switching to -15 if this is also the minimum in simulation (after confirmation?) and keeping the conversion in the track word class.

Comment (Chriisbrown) [reply DataFormats/L1TrackTrigger/interface/TTTrack_TrackWord.h:1]: I think we can put this conversion within the track quality package as a function then it can be called from anywhere it is used. I don't think having a hard coded value like this in the trackword is ideal especially if the BDT is changed to a different precision and the track word would also need updating. @tomalin can comment better if we want to hard code things into the track word in this way when they could be derived from the precision of the BDT.

Comment (cgsavard) [reply DataFormats/L1TrackTrigger/interface/TTTrack_TrackWord.h:1]: I had a chat with him on this and he agrees that putting it in the track quality class is a better fit. The change has been made as well as the minimum bin change.

**Comment by cgsavard** [comment]
All comments have been resolved.

**Review by tomalin — APPROVED** [review]
This looks good to me thanks. Though I won't merge the PR until @Chriisbrown also approves it.

I would suggest that you add a few sentences about the Track Quality to TrackFindingTracklet/README.md.

**Comment by Chriisbrown** [comment]
I'm happy with everything so can go ahead and merge. We should also put in a PR [here](https://github.com/cms-data/L1Trigger-TrackTrigger) to remove the depreciated ONNX model

**Pull request merged**
